### PR TITLE
[DSD] Support flattening the optimizer state_dict when saving and unflattening when loading

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #127604
* #127385
* #127384
* __->__ #127071
* #127070

Fixes https://github.com/pytorch/pytorch/issues/126595

**What does this PR do?**
This PR unflattens the optimizer state_dict, similar to what TorchRec does. The current `get_optimizer_state_dict()` converts the parameter IDs to FQNs in order to avoid any conflict with different optimizers on different ranks. The current returned optimizer state_dict looks like the following one:
```
{
    "state": {
          "layer1.weight": {"step": 10, "exp_avg": SomeTensor, "exp_avg_sq": SomeTensor},
          "layer2.weight": {"step": 10, "exp_avg": SomeTensor, "exp_avg_sq": SomeTensor},
    },
    "param_group": [
         {"lr": 0.0, "betas": (0.9, 0.95), ..., "params": ["layer1.weight", "layer2.weight"]}
    ]
}
```
While this can avoid the conflict and can support merging multiple optimizers use case (e.g., optimizer in backward), the current optimizer state_dict still cannot support MPMD (e.g., pipeline parallelism). The root cause is `param_group`. `param_group` cannot generate unique keys during saving -- DCP will flatten the dict but for `param_group`, DCP will get the keys like, `param_group.lr` or `param_group.params`. These keys will conflict when using pipeline parallelism.

This PR flatten the optimizer state_dict to the one as the following one:
```
{
    "state.layer1.weight.step": 10,
    "state.layer2.weight.step": 10,
    "state.layer1.weight.exp_avg": SomeTensor,
    "state.layer2.weight.exp_avg": SomeTensor,    
    "state.layer1.weight.exp_avg_sq": SomeTensor,
    "state.layer2.weight.exp_avg_sq": SomeTensor,        
    "param_group.layer1.weight.lr" : 0.1,
    "param_group.layer2.weight.lr" : 0.1,
    "param_group.layer1.weight.betas" : (0.9, 0.95),
    "param_group.layer2.weight.betas" : (0.9, 0.95),    
}
```
This allows distributed state_dict (DSD) to support MPMD (e.g., pipeline parallelism).

**Pros and Cons**
*Pros*
1. Can support optimizer resharding (e.g., changing the parallelisms from 3D to 2D or changing the number of workers).
2. User don't need to manually add prefix to different optimizer.
3. Allow users to merge the optimizer states easily. One use case is loop-based pipeline parallelism.

*Cons*
1. The implementation has a strong assumption of the structure of `param_groups` and its value. If the assumption changes or some customized optimizers do not meet the assumption, the implementations will be broken.
2. There will be extra values saved in the checkpoints. The assumption here is `param_group` generally contains scalars which are cheap to save.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @LucasLLC

### CODE REVIEW

File: torch/distributed/checkpoint/state_dict.py [hunk 1]
```diff
@@ -563,6 +564,29 @@ def _init_optim_state(optim: torch.optim.Optimizer) -> None:
...
+    for fqn, state in state_dict[STATE].items():
+        for k, v in state.items():
+            _raise_if_type_not_supported(v)
+            ret[f"{STATE}.{fqn}.{k}"] = v
+
+    for param_group in state_dict[PG]:
+        fqns = param_group.pop(PARAMS)
+        for fqn in fqns:
+            for k, v in param_group.items():
+                ret[f"{PG}.{fqn}.{k}"] = v
```
Comment (wconstab) [inline torch/distributed/checkpoint/state_dict.py:586]: it looks like this flattening would not preserve the original param-group groupings.  It is a true flattening.  Is that correct?

I'm not sure if that's good or bad- does that mean that we make the optimizer states generic enough that resharing would be possible?  (But I'm not sure how we'd reconstruct the original param_group groupings on load side.)

Comment (wconstab) [reply torch/distributed/checkpoint/state_dict.py:586]: ok reading the 'unflattening' side again it makes sense.  IIUC, because the user would have created the new optimizer based on the new model/parallelisms, it may or may not have the same param_groupings as before.  But since we normalized all of the individual states to fqn, as long as they have the same FQNs (or less FQNs) in their new model, they can find the states and then group them according to the new optimizer's groupings.

So this appears to support resharding?

**Comment by wconstab** [comment]
A general comment- it seems like the amount of optimizer-specific knowledge we have to bake in here is small- namely that there is a concept of ordered lists of parameter groupings and a particular schema for it.  This seems probably ok to me, even though we'd rather just have a clean dict from the optimizer in the first place and keep this code simpler.

I'm assuming this is true for all of our optimizers though. If the optimizers vary and not all of them have param groupings, then this could be fragile.

**Comment by fegin** [comment]
> A general comment- it seems like the amount of optimizer-specific knowledge we have to bake in here is small- namely that there is a concept of ordered lists of parameter groupings and a particular schema for it. This seems probably ok to me, even though we'd rather just have a clean dict from the optimizer in the first place and keep this code simpler.
> 
> I'm assuming this is true for all of our optimizers though. If the optimizers vary and not all of them have param groupings, then this could be fragile.

Yes, the implementation is based on all the built-in optimizers, SGD, Adam, Adamw, but should also support optimizers from Apex. However, one thing I'm not so sure is that if we will have issues with second-order optimizers. My guess is no as the second-order optimizers generally have more complicated states but the param group structure is the same. But I don't know if this is true for all the optimizers. cc., @wz337

File: torch/distributed/checkpoint/state_dict.py (hunk 1 above)
Comment (fegin) [reply torch/distributed/checkpoint/state_dict.py:586]: This supports resharding.

File: torch/distributed/checkpoint/state_dict.py [hunk 2]
```diff
@@ -563,6 +564,29 @@ def _init_optim_state(optim: torch.optim.Optimizer) -> None:
...
+    ret: Dict[str, ValueType] = {}
+    for fqn, state in state_dict[STATE].items():
+        for k, v in state.items():
+            _raise_if_type_not_supported(v)
+            ret[f"{STATE}.{fqn}.{k}"] = v
+
+    for param_group in state_dict[PG]:
+        fqns = param_group.pop(PARAMS)
+        for fqn in fqns:
+            for k, v in param_group.items():
```
[the remark this answers is not in the document]
Comment (fegin) [reply torch/distributed/checkpoint/state_dict.py:585]: Yes, it is guaranteed. But we can definitely add an assert here.

**Review by wconstab — APPROVED** [review]
lgtm. i dont think its too risky, but happy to let other DCP/optimizer experts weigh in

File: torch/distributed/checkpoint/state_dict.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (janeyx99) [reply torch/distributed/checkpoint/state_dict.py:626]: This is guaranteed today, we have test cases for it.

**Review by wz337 — APPROVED** [review]
Aligned offline about the solution. LGTM.
